Fix lru_cache where config is used

nicholasw-gc opened the pull request:
Ensure that any use of functools.lru_cache does not prevent config from being changed after the function has already run.


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec @XilunWu

nicholasw-gc commented:
@eellison how is it going? Any chance of getting the workflows approved as that runs a few more tests? Many thanks!

eellison commented:
Hey was on pto. cc @masnesral seems related to both process pool changes and the lru_cache invalidation you did.. want to take a look ?

masnesral commented:
@nicholasw-gc can you give some more context on the motivation here? At least one of these settings (is_fbcode) probably doesn't make sense to allow changing (even though its presence in this file makes it appear otherwise). The subprocess pool-related settings are also questionable because I believe it would cause the spawning of new pools without cleaning up an existing pool. Probably the others are ok.

nicholasw-gc commented:
We came across unexpected behaviour with a use of lru_cache leading to a full review across the codebase. I agree that the fbcode probably does not need a change on the basis that the setting would probably not change: it just looks wrong in isolation but that can also be addressed with a comment saying "This will not change during runtime".  

Is ` thread_pool.shutdown(wait=True, cancel_futures=False)` insufficient clearup?  Hopefully anything left is just memory which would be garbage collected.

masnesral commented:
@nicholasw-gc so sorry for the delay here. Got hung up with pytorch conference, internal oncall, and a few other excuses. I was actually more concerned with the process pool vs. the thread pool. I experimented a little with your impl and it seems that if we modify the `worker_start_method` setting between calls to `torch.compile`, then we get subprocesses hanging around that don't go away until the process exits. Similarly, if we change the `compile_threads` (which also controls the # of subprocesses in the subproc pool), we get a bunch of separate subprocess pools hanging around. So at least as implemented currently, I don't think it makes a lot of sense to give the illusion that changing these settings after initialization is robust. Or if we do want to support that, we should probably fix up the subproc pool management code to do "the right thing" and tear down the previous pool gracefully when creating a new one.

nicholasw-gc commented:
@masnesral Thanks for the response. I've also been off two weeks. 

Sorry I neglected to spot the different behaviour of process_pool_impl compared to the pool_impl. 

I think there are two possible fixes then. The first is simply to prevent the config values from being changed once read, for threading only. Would this work?
 
The second fix is to wait upon the pool being shutdown as in the thread pool. However, I am concerned that you appear to be suggesting that the subprocesses never close until the parent process closes and so a shutdown will hang. (NB the thread pool might have the same issue). Is that correct? 

A third option, of course, is to leave the current code be, i.e. effectively the user can change the config values but nothing happens.

Which would you prefer?

masnesral commented:
> The first is simply to prevent the config values from being changed once read, for threading only. Would this work?

yeah, I think it would be ok to disallow changing after they're initialized, I'm just not sure how you'd propose to do it. Is there some creative Python magic that you know about? I spent a little time investigating whether we could build some kind of assign-once class to use here, but it seems the logic that examines this file to build the config data structure only supports simple types (https://github.com/pytorch/pytorch/blob/5fb30df7d6ecc25cc7c4c17a8a33d14ddaa7c279/torch/_inductor/config.py#L1276). So it was not going to be straightforward to implement something like that. It might have been nice if changing these values were hidden behind an API call, but I don't think we can change that now :/

> The second fix is to wait upon the pool being shutdown as in the thread pool. However, I am concerned that you appear to be suggesting that the subprocesses never close until the parent process closes and so a shutdown will hang.

There might be some hang possible here, but that's not the behavior I saw. I just meant there's no functionality to gracefully shutdown (or resize) any existing subprocess pool when we create a new one due to config values changing. You can cause this to happen by changing the values of `compile_threads` or `worker_start_method` in between calls to `torch.compile(fn)(args)`, for example. That might be an unusual usage, but I have internal reports where a proliferation of subprocesses is already a problem.

> A third option, of course, is to leave the current code be, i.e. effectively the user can change the config values but nothing happens.

Well, hmmm.... Now that you mention it, I think I'll try to remove `worker_start_method` some time in the next few weeks or months since one of the two options ("fork") is problematic anyway and I don't know of a reason to keep it. `compile_threads` will probably need to stay though. While I think about that one, lemme hear your thoughts on the above.

nicholasw-gc commented:
> > The first is simply to prevent the config values from being changed once read, for threading only. Would this work?
> 
> yeah, I think it would be ok to disallow changing after they're initialized, I'm just not sure how you'd propose to do it. Is there some creative Python magic that you know about? I spent a little time investigating whether we could build some kind of assign-once class to use here, but it seems the logic that examines this file to build the config data structure only supports simple types (
> 
> https://github.com/pytorch/pytorch/blob/5fb30df7d6ecc25cc7c4c17a8a33d14ddaa7c279/torch/_inductor/config.py#L1276
> 
> ). So it was not going to be straightforward to implement something like that. It might have been nice if changing these values were hidden behind an API call, but I don't think we can change that now :/

Fortunately it turns that the `install_config_module` you linked to already has Python magic turning it into a class and so uses `__setattr__` as the API. Initially, I thought about having a decorator (or even a faux decorator in comments which is used elsewhere) but then realised it might be easier just to let anything using the config call `config.freeze_key(...)` .

```
diff --git a/test/inductor/test_config.py b/test/inductor/test_config.py
index 15bafc5233e..072d8f4d77f 100644
--- a/test/inductor/test_config.py
+++ b/test/inductor/test_config.py
@@ -57,6 +57,17 @@ class TestInductorConfig(TestCase):
         self.assertEqual(config.max_fusion_size, 321)
         self.assertEqual(config.triton.cudagraphs, False)
 
+    def test_freeze(self):
+        config.compile_threads = 1
+        config.compile_threads = 2
+        config.freeze_key("compile_threads")
+
+        def change_compile_threads():
+            config.compile_threads = 3
+
+        self.assertRaises(AttributeError, change_compile_threads)
+        config.compile_threads = 4
+
     def test_hasattr(self):
         self.assertTrue(hasattr(config, "max_fusion_size"))
         self.assertFalse(hasattr(config, "missing_name"))
diff --git a/torch/utils/_config_module.py b/torch/utils/_config_module.py
index 8aa9d4063f0..9bf9fc9d19f 100644
--- a/torch/utils/_config_module.py
+++ b/torch/utils/_config_module.py
@@ -68,6 +68,7 @@ def install_config_module(module: ModuleType) -> None:
     module._config = config  # type: ignore[attr-defined]
     module._default = default  # type: ignore[attr-defined]
     module._allowed_keys = set(config.keys())  # type: ignore[attr-defined]
+    module._frozen_keys = set()  # type: ignore[attr-defined]
     module._compile_ignored_keys = compile_ignored_keys  # type: ignore[attr-defined]
     module.__class__ = ConfigModuleInstance
     module._is_dirty = True  # type: ignore[attr-defined]
@@ -129,6 +130,7 @@ class ConfigModule(ModuleType):
     _allowed_keys: Set[str]
     _bypass_keys: Set[str]
     _compile_ignored_keys: Set[str]
+    _frozen_keys: Set[str]
     _is_dirty: bool
     _hash_digest: Optional[bytes]
 
@@ -142,6 +144,13 @@ class ConfigModule(ModuleType):
             super().__setattr__(name, value)
         elif name not in self._allowed_keys:
             raise AttributeError(f"{self.__name__}.{name} does not exist")
+        elif name in self._frozen_keys:
+            raise AttributeError(
+                f"{self.__name__}.{name} is frozen so you cannot "
+                "change its value. Consider setting the value "
+                "before using any functionaliy which relies on "
+                "the value"
+            )
         else:
             self._config[name] = value
 
@@ -177,6 +186,12 @@ class ConfigModule(ModuleType):
             config[key] = self._config[key]
         return config
 
+    def freeze_key(self, name: str) -> None:
+        """Freeze a key to prevent the value from being changed by the user.
+        Use this if a component has already read the value in the config such
+        that changing the value will have no effect."""
+        self._frozen_keys.add(name)
+
     def codegen_config(self) -> str:
         """Convert config to Python statements that replicate current config.
         This does NOT include config settings that are at default values.
```

nicholasw-gc commented:
> > The second fix is to wait upon the pool being shutdown as in the thread pool. However, I am concerned that you appear to be suggesting that the subprocesses never close until the parent process closes and so a shutdown will hang.
> 
> There might be some hang possible here, but that's not the behavior I saw. I just meant there's no functionality to gracefully shutdown (or resize) any existing subprocess pool when we create a new one due to config values changing. You can cause this to happen by changing the values of `compile_threads` or `worker_start_method` in between calls to `torch.compile(fn)(args)`, for example. That might be an unusual usage, but I have internal reports where a proliferation of subprocesses is already a problem.

Probably best not try and change this for now then. I'll rebase the PR.

nicholasw-gc commented:
@pytorchbot label "topic: not user facing"

masnesral commented:
@nicholasw-gc 

> Probably best not try and change this for now then. I'll rebase the PR.

Does that mean you're going to punt on changing anything related to `compile_threads` and `worker_start_method`? If so, sounds good. I'll remove `worker_start_method` anyway and take an action item on compile_threads. But if that's the plan, I still see changes in the PR that we can probably abandon, e.g., pulling out `pool_impl` and `process_pool_impl`. Lemme know if I misunderstand though.

Looks like you have some merge markers after rebasing.

For is_fbcode(), I think that's no longer a config option since you first made this PR (got replaced by a function). So I think you can ignore that too.

nicholasw-gc commented:
> @nicholasw-gc
> 
> > Probably best not try and change this for now then. I'll rebase the PR.
> 
> Does that mean you're going to punt on changing anything related to `compile_threads` and `worker_start_method`? If so, sounds good. I'll remove `worker_start_method` anyway and take an action item on compile_threads. But if that's the plan, I still see changes in the PR that we can probably abandon, e.g., pulling out `pool_impl` and `process_pool_impl`. Lemme know if I misunderstand though.

Just freezing `compile_threads` as it sounds like it's still needed. The pool_impl change is probably unlikely to be relevant now (likely process_pool_impl will have frozen it around the same time) so should we keep it in case it's good in the future or just freeze compile_threads in def pool()? 

> Looks like you have some merge markers after rebasing
Oops I forgot to "git add -u" so the commit I pushed was out of date. 

> For is_fbcode(), I think that's no longer a config option since you first made this PR (got replaced by a function). So I think you can ignore that too.
Did you mean the is_fbcode changes can be removed?

nicholasw-gc commented:
> > or just freeze compile_threads in def pool()?
> 
> I see. I didn't look closely enough at the diff you posted inline; I see how it works now. Maybe put the freezing in `get_compile_threads()` would give us what we want then? 

I did not do this because of the comment ` Temporary for internal rollout. Assign config.compile_threads lazily and return it. TODO: remove after rollout.`
It's not clear to me how long Temporary is but it sounded like it would be matter of days to weeks before it' removed.

>Also, if we do the freezing thing can we make sure we don't throw if we're attempting to set a value that's the same as the existing value? I'm a little skittish about introducing new exceptions and breaking existing usages (albeit usage that don't work as expected). At least we can ignore noops?

I have changed this in the latest push. 


> > Did you mean the is_fbcode changes can be removed?
> 
> Yes, I believe it's been changed to a function call now -- it's no longer a config option

Should also all be changed too. :)

masnesral commented on test/inductor/test_config.py:
```diff
@@ -57,6 +58,41 @@ def test_save_load(self):
+        self.assertRaises(AttributeError, change_compile_threads)
+
+    def test_load_save_freeze_same(self):
+        # Loading the same value is akways fine
```
akways -> always

nicholasw-gc replied:
Cheers!

masnesral commented on torch/utils/_config_module.py:
```diff
@@ -177,6 +190,12 @@ def save_config_portable(self) -> Dict[str, Any]:
+        """Freeze a key to prevent the value from being changed by the user.
+        Use this if a component has already read the value in the config such
+        that changing the value will have no effect."""
+        self._frozen_keys.add(name)
```
Shall we check if the name is valid?

nicholasw-gc replied:
Done :)

masnesral commented on torch/utils/_config_module.py:
```diff
@@ -177,6 +190,12 @@ def save_config_portable(self) -> Dict[str, Any]:
             config[key] = self._config[key]
         return config
 
+    def freeze_key(self, name: str) -> None:
```
@eellison do you see any problem with this? TL;DR this PR has new functionality to "freeze" config settings that we don't support changing after they're first read. This is how that's accomplished.

masnesral commented:
@nicholasw-gc hmm, the test failure(s) looks real, but not immediately clear to me what the problem is.

nicholasw-gc commented:
All three start with "AttributeError: torch._inductor.config.compile_threads is frozen so you cannot change its value. Consider setting the value before using any functionality which freezes the relevant config value"

But it's not clear to me why that yields another exception. I reckon that is just how unitest.mock reports any exception it gets no matter what the real cause is.

nicholasw-gc commented:
> @nicholasw-gc hmm, the test failure(s) looks real, but not immediately clear to me what the problem is.
Ok I changed a lot now, essentially tests which changed `compile_threads` which was presumably having no effect (except for changing to 1 which bypasses the pool) unless the test was run in isolation. I have not tested any specialist hardware tests, so it would be best to approve workflows but wait for all tests to see what happens.

masnesral commented:
@nicholasw-gc I feel bad because clearly there's a lot of thought in this last round of changes, but after seeing what's required I'm feeling like all the added complexity is not warranted just to support freezing for `compile_threads`. I suggest we go back to what you had before, but:
1) For `worker_start_method`, it's fine with me if we freeze the config option or just leave it "unsafe" w.r.t the lru_cache. I'll be removing this config option shortly anyway.
2) For `compile_threads`, I think it's fine if we don't support resizing pools when the config option changes after first read. I think the important part to support is whether it changes between 1 and not 1 since that distinction controls whether we attempt to compile in parallel or just do it sequentially in the main process. For example: https://github.com/pytorch/pytorch/blob/a0a978ce237a84e9da66a36924c6fa670f9c19bc/torch/_inductor/async_compile.py#L199. If you want to make a change to achieve that in this PR, sounds great. I might have missed something, but I think that route means we need almost no changes. But also If you're fed up at this point, I don't mind taking an action item to clean up the handling of 'compile_threads`.

Sorry for the churn; what do you think?

nicholasw-gc commented:
> @nicholasw-gc I feel bad because clearly there's a lot of thought in this last round of changes, but after seeing what's required I'm feeling like all the added complexity is not warranted just to support freezing for `compile_threads`. I suggest we go back to what you had before, but:
> 
> 1. For `worker_start_method`, it's fine with me if we freeze the config option or just leave it "unsafe" w.r.t the lru_cache. I'll be removing this config option shortly anyway.
> 2. For `compile_threads`, I think it's fine if we don't support resizing pools when the config option changes after first read. I think the important part to support is whether it changes between 1 and not 1 since that distinction controls whether we attempt to compile in parallel or just do it sequentially in the main process. For example: https://github.com/pytorch/pytorch/blob/a0a978ce237a84e9da66a36924c6fa670f9c19bc/torch/_inductor/async_compile.py#L199
>    . If you want to make a change to achieve that in this PR, sounds great. I might have missed something, but I think that route means we need almost no changes. But also If you're fed up at this point, I don't mind taking an action item to clean up the handling of 'compile_threads`.
> 
> Sorry for the churn; what do you think?

How about this PR becomes fixes apart from _compile_threads_, which is intentionally left "wrong" (save for the 1 vs not one of course)?  That way we can merge it before too much changes.

The rest of the changes, would you be keen to take the remainder as a diff and incorporate it as part of any changes including `worker_start_method`.

masnesral commented:
@nicholasw-gc Sounds perfect. Thanks for your patience working through this with me.

nicholasw-gc committed:
Fix lru_cache where config is used

- Ensure that any use of functools.lru_cache does not prevent config
  from being changed after the function has already run.
- Exclude "compile_threads" due to complications: changing to and
  from "1" will have the intented effect but other values will
  silently be ignored.

nicholasw-gc commented:
> @nicholasw-gc Sounds perfect. Thanks for your patience working through this with me.

Should work now. I've loaded what's been removed.

[Uploading compile_threads_diff.patch…]()

nicholasw-gc commented:
@masnesral good to run workflows and merge?

nicholasw-gc commented:
> @nicholasw-gc, sorry this is still not what I was expecting given the last conversation. If we're going to leave `compile_threads` alone,
> 
> > How about this PR becomes fixes apart from compile_threads, which is intentionally left "wrong"
> 
> then I don't understand why we need all the changes to the test files, e.g., `@patch.object(torch._inductor.config, "do_not_use_compile_threads", True)`.

Sorry I separated them into two commits, then I managed to push the branch with **both** commits instead of just the one. :(

Should be correct now

masnesral commented:
Yep, thanks. Looks good.

nicholasw-gc commented:
@pytorchbot merge

nicholasw-gc commented:
@pytorchbot merge

nicholasw-gc commented:
@pytorchbot merge

nicholasw-gc commented:
@pytorchbot merge
